<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Alignment - Andersen Lab Dry Guide</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Alignment";
    var mkdocs_page_input_path = "pipeline-alignment.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Andersen Lab Dry Guide</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../bash/">Bash</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Quest</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quest-intro/">Introduction</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-andersen-lab-env/">Andersen-Lab-Env</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-nextflow/">Nextflow</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-mount/">Mounting Quest</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Pipelines</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../pipeline-overview/">Overview</a>
                </li>
                <li class="">
                    
    <a class="" href="../adding-seq-data/">Adding NILs Sequencing Data</a>
                </li>
                <li class="">
                    
    <a class="" href="../sample-sheets/">Sample Sheets</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-GCPconfig/">GCP configuration</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-docker/">Build docker iamge</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-trimming/">Trimming</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-nil-ril/">NIL / RIL</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">Alignment</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#alignment-nf">alignment-nf</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#usage">Usage</a></li>
        
            <li><a class="toctree-l4" href="#pipeline_overview">Pipeline Overview</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#usage_1">Usage</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#quick_start">Quick Start</a></li>
        
            <li><a class="toctree-l4" href="#profiles_and_running_the_pipeline">Profiles and Running the Pipeline</a></li>
        
            <li><a class="toctree-l4" href="#software">Software</a></li>
        
            <li><a class="toctree-l4" href="#debugging">Debugging</a></li>
        
            <li><a class="toctree-l4" href="#--output">--output</a></li>
        
            <li><a class="toctree-l4" href="#--fqs_fastqs">--fqs (FASTQs)</a></li>
        
            <li><a class="toctree-l4" href="#--fqs_file_prefix">--fqs_file_prefix</a></li>
        
            <li><a class="toctree-l4" href="#--reference">--reference</a></li>
        
            <li><a class="toctree-l4" href="#--email">--email</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#sample_sheet">Sample Sheet</a></li>
    

    <li class="toctree-l3"><a href="#output">Output</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#strain">Strain</a></li>
        
            <li><a class="toctree-l4" href="#isotype">Isotype</a></li>
        
        </ul>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-concordance/">Concordance</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-wi/">WI</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-cegwas/">Cegwas</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../cendr/">CeNDR</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../r/">R</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../backup/">Backup</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../cloud/">Cloud</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../labsite/">Andersen Labsite</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Andersen Lab Dry Guide</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Pipelines &raquo;</li>
        
      
    
    <li>Alignment</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/Andersenlab/dry-guide/edit/master/docs/pipeline-alignment.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="alignment-nf">alignment-nf<a class="headerlink" href="#alignment-nf" title="Permanent link">&para;</a></h1>
<p>The <a href="https://github.com/AndersenLab/alignment-nf">alignment-nf</a> pipeline performs alignment for wild isolate sequence data <strong>at the strain level</strong>, and outputs BAMs and related information. Those BAMs can be used for downstream analysis of <a href="http://andersenlab.org/dry-guide/pipeline-concordance/">concordance-nf</a>, <a href="http://andersenlab.org/dry-guide/pipeline-wi/">wi-nf</a> and variants calling.</p>
<p>Note that historically, sequence processing was previously performed at the isotype level. We are still interested in filtering strains used in analysis at the isotype level, but alignment and variant calling are now performed at the strain level rather than at the isotype level.</p>
<div class="toc">
<ul>
<li><a href="#alignment-nf">alignment-nf</a><ul>
<li><a href="#usage">Usage</a></li>
<li><a href="#pipeline_overview">Pipeline Overview</a></li>
</ul>
</li>
<li><a href="#usage_1">Usage</a><ul>
<li><a href="#quick_start">Quick Start</a></li>
<li><a href="#profiles_and_running_the_pipeline">Profiles and Running the Pipeline</a></li>
<li><a href="#software">Software</a></li>
<li><a href="#debugging">Debugging</a></li>
<li><a href="#--output">--output</a></li>
<li><a href="#--fqs_fastqs">--fqs (FASTQs)</a></li>
<li><a href="#--fqs_file_prefix">--fqs_file_prefix</a></li>
<li><a href="#--reference">--reference</a></li>
<li><a href="#--email">--email</a></li>
</ul>
</li>
<li><a href="#sample_sheet">Sample Sheet</a></li>
<li><a href="#output">Output</a><ul>
<li><a href="#strain">Strain</a></li>
<li><a href="#isotype">Isotype</a></li>
</ul>
</li>
</ul>
</div>
<h2 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link">&para;</a></h2>
<pre><code>             ▗▖ ▝▜   ▝                       ▗      ▗▖ ▖▗▄▄▖
             ▐▌  ▐  ▗▄   ▄▄ ▗▗▖ ▗▄▄  ▄▖ ▗▗▖ ▗▟▄     ▐▚ ▌▐
             ▌▐  ▐   ▐  ▐▘▜ ▐▘▐ ▐▐▐ ▐▘▐ ▐▘▐  ▐      ▐▐▖▌▐▄▄▖
             ▙▟  ▐   ▐  ▐ ▐ ▐ ▐ ▐▐▐ ▐▀▀ ▐ ▐  ▐   ▀▘ ▐ ▌▌▐
            ▐  ▌ ▝▄ ▗▟▄ ▝▙▜ ▐ ▐ ▐▐▐ ▝▙▞ ▐ ▐  ▝▄     ▐ ▐▌▐
                         ▖▐
                         ▝▘
    parameters              description                    Set/Default
    ==========              ===========                    ========================
    --debug                 Set to 'true' to test          ${params.debug}
    --fqs                   fastq file (see help)          ${params.fqs}
    --fq_file_prefix        fastq prefix                   ${params.fq_file_prefix}
    --reference             Reference Genome (w/ .gz)      ${params.reference}
    --output                Location for output            ${params.output}
    --email                 Email to be sent results       ${params.email}

    HELP: http://andersenlab.org/dry-guide/pipeline-alignment/
</code></pre>

<h2 id="pipeline_overview">Pipeline Overview<a class="headerlink" href="#pipeline_overview" title="Permanent link">&para;</a></h2>
<p><img alt="Pipeline-overview" src="../img/alignment-nf.png" /></p>
<h1 id="usage_1">Usage<a class="headerlink" href="#usage_1" title="Permanent link">&para;</a></h1>
<h2 id="quick_start">Quick Start<a class="headerlink" href="#quick_start" title="Permanent link">&para;</a></h2>
<p>The pipeline can be run using the following command:</p>
<pre><code>NXF_VER=19.09.0-edge nextflow run main.nf -profile local
</code></pre>

<h2 id="profiles_and_running_the_pipeline">Profiles and Running the Pipeline<a class="headerlink" href="#profiles_and_running_the_pipeline" title="Permanent link">&para;</a></h2>
<p>There are three configuration profiles for this pipeline.</p>
<ul>
<li><code>local</code> - Used for local development.</li>
<li><code>quest</code> - Used for running on Quest.</li>
<li><code>gcp</code> - For running on Google Cloud.</li>
</ul>
<h2 id="software">Software<a class="headerlink" href="#software" title="Permanent link">&para;</a></h2>
<p>Almost all processes within the pipeline are now managed using <a href="">conda</a>. To use the pipeline, you must have <code>conda</code> installed and available. Nextflow will take care of installing conda environments and managing software.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a href="https://www.github.com/brentp/mosdepth">mosdepth</a> is used to calculate coverage. <code>mosdepth</code> is available on Linux machines, but not on Mac OSX. That is why the conda environment for the <code>coverage</code> process is specified as <code>conda { System.properties['os.name'] != "Mac OS X" ? 'bioconda::mosdepth=0.2.6' : "" }</code>. This snippet allows mosdepth to run off the version present in the <code>bin</code> folder locally on Mac OSX, or using the conda-managed installation on a Linux Machine.</p>
</div>
<h2 id="debugging">Debugging<a class="headerlink" href="#debugging" title="Permanent link">&para;</a></h2>
<p>You should use <code>--debug true</code> for testing/debugging purposes. This will run the debug test set using your specified config.</p>
<p>This can be used locally or on quest. For example:</p>
<pre><code>nextflow run main.nf -profile quest_debug -resume --goal &quot;strain&quot;
</code></pre>

<h2 id="--output">--output<a class="headerlink" href="#--output" title="Permanent link">&para;</a></h2>
<p>A directory in which to output results. By default it will be <code>alignment-YYYY-MM-DD</code> where YYYY-MM-DD is todays date. If you have set <code>--debug true</code>, it will take the form of <code>alignment-YYYY-MM-DD-debug</code>.</p>
<h2 id="--fqs_fastqs">--fqs (FASTQs)<a class="headerlink" href="#--fqs_fastqs" title="Permanent link">&para;</a></h2>
<p>When running the <code>alignment-nf</code> pipeline you must provide a sample sheet that tells it where fastqs are and which samples group into strains/isotypes. By default, this is the sample sheet in the base of the alignment-nf repo , but can be specified using <code>--fqs</code> if an alternative is required. The sample sheet provides information on the strain/isotype, fastq_pairs, library, location of fastqs, and sequencing folder. See <a href="# Sample Sheet">Sample Sheet</a> for more details on how it is generated.</p>
<p>When using <code>--debug true</code>, the <code>test_data/sample_sheet.tsv</code> file is used.</p>
<h2 id="--fqs_file_prefix">--fqs_file_prefix<a class="headerlink" href="#--fqs_file_prefix" title="Permanent link">&para;</a></h2>
<p>A prefix path for FASTQs defined in a sample sheet. The sample sheet designed for usage on Quest (<code>SM_sample_sheet.tsv</code>) uses absolute paths so no FASTQ prefix is necessary. Additionally, there is no need to set this option as it is set for you when using the <code>-profile</code> flag. This option is only necessary (maybe) with a custom dataset where you are not using absolute paths to reference FASTQs.</p>
<h2 id="--reference">--reference<a class="headerlink" href="#--reference" title="Permanent link">&para;</a></h2>
<p>A fasta reference indexed with BWA. This is packaged with the repo for convenience for running locally.</p>
<p>On Quest, the reference is available here:</p>
<pre><code>/projects/b1059/data/genomes/c_elegans/WS245/WS245.fa.gz
</code></pre>

<h2 id="--email">--email<a class="headerlink" href="#--email" title="Permanent link">&para;</a></h2>
<p>Enable a email notification when you use this params.</p>
<h1 id="sample_sheet">Sample Sheet<a class="headerlink" href="#sample_sheet" title="Permanent link">&para;</a></h1>
<p>A sample sheet is the only acceptable input style for sequences in this pipeline. Sample sheet should be constructed by using the scripts under <code>scripts</code> folder, or follow the format in <a href="http://andersenlab.org/dry-guide/sample-sheets/">Sample Sheets</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you don't use <code>--fqs</code> to specify the directory of sample sheet, you should put sample sheet in the root of this repo. The name structure is restricted as <code>SM_strain_sheet.tsv</code> or <code>SM_isotype_sheet.tsv</code>.</p>
</div>
<h1 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h1>
<h2 id="strain">Strain<a class="headerlink" href="#strain" title="Permanent link">&para;</a></h2>
<p>When you use <code>--goal "strain"</code></p>
<pre><code>├── BAM
│   ├── WN2065.bam
│   └── WN2065.bam.bai
├── fq
│   ├── fq_bam_idxstats.tsv
│   ├── fq_bam_stats.tsv
│   ├── fq_coverage.full.tsv
│   └── fq_coverage.tsv
├── log.txt
└── SM
    ├── bam_duplicates.tsv
    ├── SM_bam_idxstats.tsv
    ├── SM_bam_stats.tsv
    ├── SM_coverage.full.tsv
    ├── SM_coverage.mb.tsv.gz
    └── SM_coverage.tsv
</code></pre>

<ul>
<li><strong>log.txt</strong> - A summary of nextflow run.</li>
<li><strong>BAM</strong> - BAMs for strains are included in this folder.</li>
<li><strong>fq</strong> - The coverage and BAMs information for each sample were included here.</li>
<li><strong>SM</strong> - The coverage and BAMs information for each strain, samples come from the same strain were merged and presented here.</li>
</ul>
<h2 id="isotype">Isotype<a class="headerlink" href="#isotype" title="Permanent link">&para;</a></h2>
<p>When you use <code>--goal "isotype"</code></p>
<pre><code>├── BAM
│   ├── RC301.bam
│   └── RC301.bam.bai
├── log.txt
├── phenotype
│   ├── MT_content.tsv
│   └── telseq.tsv
├── report
│   ├── multiqc_data
│   │   ├── multiqc_data.json
│   │   ├── multiqc_fastqc.json
│   │   ├── multiqc_general_stats.json
│   │   ├── multiqc_picard_AlignmentSummaryMetrics.json
│   │   ├── multiqc_picard_insertSize.json
│   │   ├── multiqc_samtools_idxstats.json
│   │   ├── multiqc_samtools_stats.json
│   │   └── multiqc_sources.json
│   └── multiqc.html
└── SM
    ├── isotype_bam_idxstats.tsv
    ├── isotype_bam_stats.tsv
    ├── isotype_coverage.full.tsv
    └── isotype_coverage.tsv
</code></pre>

<ul>
<li><strong>log.txt</strong> - A summary of nextflow run.</li>
<li><strong>BAM</strong> - BAMs for isotype are included in this folder.</li>
<li><strong>phenotype</strong> - Mitochondrial content and Telomere length for each isotype.</li>
<li><strong>SM</strong> - Alignment statistics by isotype, samples come from the same isotype were merged.</li>
<li><strong>report</strong> - A comprehensive report of isotype BAMs.</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../pipeline-concordance/" class="btn btn-neutral float-right" title="Concordance">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../pipeline-nil-ril/" class="btn btn-neutral" title="NIL / RIL"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/Andersenlab/dry-guide/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../pipeline-nil-ril/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../pipeline-concordance/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
