<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Alignment - Andersen Lab Dry Guide</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Alignment";
    var mkdocs_page_input_path = "pipeline-alignment.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Andersen Lab Dry Guide</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../bash/">Bash</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Quest</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../quest-intro/">Introduction</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-andersen-lab-env/">Andersen-Lab-Env</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-nextflow/">Nextflow</a>
                </li>
                <li class="">
                    
    <a class="" href="../quest-mount/">Mounting Quest</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Pipelines</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../pipeline-overview/">Overview</a>
                </li>
                <li class="">
                    
    <a class="" href="../adding-seq-data/">Adding NILs Sequencing Data</a>
                </li>
                <li class="">
                    
    <a class="" href="../sample-sheets/">Sample Sheets</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-GCPconfig/">GCP configuration</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-docker/">Build docker iamge</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-trimming/">Trimming</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-nil-ril/">NIL / RIL</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">Alignment</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#alignment-nf">alignment-nf</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#usage">Usage</a></li>
        
            <li><a class="toctree-l4" href="#pipeline_overview">Pipeline Overview</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#usage_1">Usage</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#quick_start">Quick Start</a></li>
        
            <li><a class="toctree-l4" href="#profiles_and_running_the_pipeline">Profiles and Running the Pipeline</a></li>
        
            <li><a class="toctree-l4" href="#software">Software</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#parameters">Parameters</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#--debug">--debug</a></li>
        
            <li><a class="toctree-l4" href="#--sample_sheet">--sample_sheet</a></li>
        
            <li><a class="toctree-l4" href="#--fq_prefix">--fq_prefix</a></li>
        
            <li><a class="toctree-l4" href="#--kmers">--kmers</a></li>
        
            <li><a class="toctree-l4" href="#--reference">--reference</a></li>
        
            <li><a class="toctree-l4" href="#--output">--output</a></li>
        
            <li><a class="toctree-l4" href="#--email">--email</a></li>
        
        </ul>
    

    <li class="toctree-l3"><a href="#output">Output</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#strain">Strain</a></li>
        
        </ul>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-concordance/">Concordance</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-wi/">WI</a>
                </li>
                <li class="">
                    
    <a class="" href="../pipeline-cegwas/">Cegwas</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../cendr/">CeNDR</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../r/">R</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../backup/">Backup</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../cloud/">Cloud</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../labsite/">Andersen Labsite</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Andersen Lab Dry Guide</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Pipelines &raquo;</li>
        
      
    
    <li>Alignment</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/Andersenlab/dry-guide/edit/master/docs/pipeline-alignment.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="alignment-nf">alignment-nf<a class="headerlink" href="#alignment-nf" title="Permanent link">&para;</a></h1>
<p>The <a href="https://github.com/AndersenLab/alignment-nf">alignment-nf</a> pipeline performs alignment for wild isolate sequence data <strong>at the strain level</strong>, and outputs BAMs and related information. Those BAMs can be used for downstream analysis including variant calling, <a href="http://andersenlab.org/dry-guide/pipeline-concordance/">concordance analysis</a>, <a href="http://andersenlab.org/dry-guide/pipeline-wi/">wi-nf (variant calling)</a> and other analyses.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Historically, sequence processing was performed at the isotype level. We are still interested in filtering strains used in analysis at the isotype level, but alignment and variant calling are now performed at the strain level rather than at the isotype level.</p>
</div>
<div class="toc">
<ul>
<li><a href="#alignment-nf">alignment-nf</a><ul>
<li><a href="#usage">Usage</a></li>
<li><a href="#pipeline_overview">Pipeline Overview</a></li>
</ul>
</li>
<li><a href="#usage_1">Usage</a><ul>
<li><a href="#quick_start">Quick Start</a></li>
<li><a href="#profiles_and_running_the_pipeline">Profiles and Running the Pipeline</a></li>
<li><a href="#software">Software</a></li>
</ul>
</li>
<li><a href="#parameters">Parameters</a><ul>
<li><a href="#--debug">--debug</a></li>
<li><a href="#--sample_sheet">--sample_sheet</a></li>
<li><a href="#--fq_prefix">--fq_prefix</a></li>
<li><a href="#--kmers">--kmers</a></li>
<li><a href="#--reference">--reference</a></li>
<li><a href="#--output">--output</a></li>
<li><a href="#--email">--email</a></li>
</ul>
</li>
<li><a href="#output">Output</a><ul>
<li><a href="#strain">Strain</a></li>
</ul>
</li>
</ul>
</div>
<h2 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link">&para;</a></h2>
<pre><code>
             ▗▖ ▝▜   ▝                       ▗      ▗▖ ▖▗▄▄▖
             ▐▌  ▐  ▗▄   ▄▄ ▗▗▖ ▗▄▄  ▄▖ ▗▗▖ ▗▟▄     ▐▚ ▌▐
             ▌▐  ▐   ▐  ▐▘▜ ▐▘▐ ▐▐▐ ▐▘▐ ▐▘▐  ▐      ▐▐▖▌▐▄▄▖
             ▙▟  ▐   ▐  ▐ ▐ ▐ ▐ ▐▐▐ ▐▀▀ ▐ ▐  ▐   ▀▘ ▐ ▌▌▐
            ▐  ▌ ▝▄ ▗▟▄ ▝▙▜ ▐ ▐ ▐▐▐ ▝▙▞ ▐ ▐  ▝▄     ▐ ▐▌▐
                         ▖▐
                         ▝▘
    parameters              description                    Set/Default
    ==========              ===========                    ========================
    --debug                 Set to 'true' to test          ${params.debug}
    --sample_sheet          sample_sheet                   ${params.sample_sheet}
    --fq_prefix             fastq prefix                   ${params.fq_prefix}
    --kmers                 count kmers                    ${params.kmers}
    --reference             Reference Genome (w/ .gz)      ${params.reference}
    --output                Location for output            ${params.output}
    --email                 Email to be sent results       ${params.email}

    HELP: http://andersenlab.org/dry-guide/pipeline-alignment/
</code></pre>

<h2 id="pipeline_overview">Pipeline Overview<a class="headerlink" href="#pipeline_overview" title="Permanent link">&para;</a></h2>
<p><img alt="Pipeline-overview" src="../img/alignment.png" /></p>
<h1 id="usage_1">Usage<a class="headerlink" href="#usage_1" title="Permanent link">&para;</a></h1>
<h2 id="quick_start">Quick Start<a class="headerlink" href="#quick_start" title="Permanent link">&para;</a></h2>
<p>The pipeline can be run using the following command:</p>
<pre><code>NXF_VER=19.09.0-edge nextflow run main.nf -profile local
</code></pre>

<h2 id="profiles_and_running_the_pipeline">Profiles and Running the Pipeline<a class="headerlink" href="#profiles_and_running_the_pipeline" title="Permanent link">&para;</a></h2>
<p>There are three configuration profiles for this pipeline.</p>
<ul>
<li><code>local</code> - Used for local development.</li>
<li><code>quest</code> - Used for running on Quest.</li>
<li><code>gcp</code> - For running on Google Cloud.</li>
</ul>
<h2 id="software">Software<a class="headerlink" href="#software" title="Permanent link">&para;</a></h2>
<p>Almost all processes within the pipeline are now managed using <a href="">conda</a>. To use the pipeline, you must have <code>conda</code> installed and available. Nextflow will take care of installing conda environments and managing software.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a href="https://www.github.com/brentp/mosdepth">mosdepth</a> is used to calculate coverage. <code>mosdepth</code> is available on Linux machines, but not on Mac OSX. That is why the conda environment for the <code>coverage</code> process is specified as <code>conda { System.properties['os.name'] != "Mac OS X" ? 'bioconda::mosdepth=0.2.6' : "" }</code>. This snippet allows mosdepth to run off the executable present in the <code>bin</code> folder locally on Mac OSX, or use the conda-based installation when on Linux.</p>
</div>
<h1 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h1>
<h2 id="--debug">--debug<a class="headerlink" href="#--debug" title="Permanent link">&para;</a></h2>
<p>You should use <code>--debug true</code> for testing/debugging purposes. This will run the debug test set (located in the <code>test_data</code> folder) using your specified configuration profile (e.g. local / quest / gcp).</p>
<p>For example:</p>
<pre><code>nextflow run main.nf -profile local --debug -resume
</code></pre>

<p>Using <code>--debug</code> will automatically set the sample sheet to <code>test_data/sample_sheet.tsv</code></p>
<h2 id="--sample_sheet">--sample_sheet<a class="headerlink" href="#--sample_sheet" title="Permanent link">&para;</a></h2>
<p>Below is a summary of the sample sheet. Please see <a href="/sample-sheets">sample_sheet</a> for more information.</p>
<p>The sample sheet specifies the FASTQs and associated metadata required for alignment. An example sample sheet is bundled with the test data (<code>test_data/sample_sheet.tsv</code>). The first few lines look like this:</p>
<table>
<thead>
<tr>
<th align="left">strain</th>
<th>reference_strain</th>
<th>isotype</th>
<th>id</th>
<th>lb</th>
<th>fq1</th>
<th>fq2</th>
<th>seq_folder</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">AB1</td>
<td>FALSE</td>
<td>ISO1</td>
<td>BGI1-RET2-AB1</td>
<td>RET2</td>
<td>BGI1-RET2-AB1-trim-1P.fq.gz</td>
<td>BGI1-RET2-AB1-trim-2P.fq.gz</td>
<td>original_wi_set</td>
<td></td>
</tr>
<tr>
<td align="left">ECA243</td>
<td>TRUE</td>
<td>ISO1</td>
<td>BGI3-RET3b-ECA243</td>
<td>RET3b</td>
<td>BGI3-RET3b-ECA243-trim-1P.fq.gz</td>
<td>BGI3-RET3b-ECA243-trim-2P.fq.gz</td>
<td>original_wi_set</td>
<td></td>
</tr>
</tbody>
</table>
<p>The only required columns for running are:</p>
<ul>
<li><strong>strain</strong> - the name of the strain</li>
<li><strong>id</strong> - A unique ID for each sequencing run.</li>
<li><strong>lb</strong> - A library ID.</li>
<li><strong>fq1</strong> - The path to FASTQ1</li>
<li><strong>fq2</strong> - The path to FASTQ2</li>
</ul>
<p>The alignment pipeline will merge multiple sequencing runs of the same strain into a single bam. However, summary output is provided at both the <code>strain</code> and <code>id</code> level. In this way, if there is a poor sequencing run it can be identified and removed from a collection of sequencing runs belonging to a strain.</p>
<p>When using <code>--debug true</code>, the <code>test_data/sample_sheet.tsv</code> file is used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Previously this option was specified using <code>--fqs</code>.</p>
</div>
<h2 id="--fq_prefix">--fq_prefix<a class="headerlink" href="#--fq_prefix" title="Permanent link">&para;</a></h2>
<p>Within a sample sheet you may specify the locations of FASTQs using an absolute directory or a relative directory. If you want to use a relative directory, you should use the <code>--fq_prefix</code> to set the path that should be prefixed to each FASTQ.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Previously, this option was <code>--fqs_file_prefix</code></p>
</div>
<h2 id="--kmers">--kmers<a class="headerlink" href="#--kmers" title="Permanent link">&para;</a></h2>
<p><strong>default</strong> = true</p>
<p>Toggles kmer-analysis</p>
<h2 id="--reference">--reference<a class="headerlink" href="#--reference" title="Permanent link">&para;</a></h2>
<p>A fasta reference indexed with BWA. WS245 is packaged with the pipeline for convenience when testing or running locally.</p>
<p>On Quest, the reference is available here:</p>
<pre><code>/projects/b1059/data/genomes/c_elegans/WS245/WS245.fa.gz
</code></pre>

<h2 id="--output">--output<a class="headerlink" href="#--output" title="Permanent link">&para;</a></h2>
<p><strong>Default</strong> - <code>Alignment-YYYYMMDD</code></p>
<p>A directory in which to output results. If you have set <code>--debug true</code>, the default output directory will be <code>alignment-YYYYMMDD-debug</code>.</p>
<h2 id="--email">--email<a class="headerlink" href="#--email" title="Permanent link">&para;</a></h2>
<p>Setting <code>--email</code> will trigger an email report following pipeline execution.</p>
<h1 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h1>
<h2 id="strain">Strain<a class="headerlink" href="#strain" title="Permanent link">&para;</a></h2>
<pre><code>├── _aggregate
│   ├── kmers.tsv
│   └── multiqc
│       ├── id_data/
│       │   ├── ... (same as strain_data/)
│       ├── id_multiqc_report.html
│       ├── strain_data/
│       │   ├── mqc_mosdepth-coverage-dist-id_1.txt
│       │   ├── mqc_mosdepth-coverage-per-contig_1.txt
│       │   ├── mqc_mosdepth-coverage-plot-id_1.txt
│       │   ├── mqc_picard_deduplication_1.txt
│       │   ├── mqc_samtools-idxstats-mapped-reads-plot_Counts.txt
│       │   ├── mqc_samtools-idxstats-mapped-reads-plot_Normalised_Counts.txt
│       │   ├── mqc_samtools_alignment_plot_1.txt
│       │   ├── multiqc.log
│       │   ├── multiqc_data.json
│       │   ├── multiqc_general_stats.txt
│       │   ├── multiqc_picard_dups.txt
│       │   ├── multiqc_qualimap_bamqc_genome_results.txt
│       │   ├── multiqc_samtools_flagstat.txt
│       │   ├── multiqc_samtools_idxstats.txt
│       │   ├── multiqc_samtools_stats.txt
│       │   └── multiqc_sources.txt
│       └── strain_multiqc_report.html
├── bam
│   ├── [strain].bam
│   └── [strain].bam.bai
├── coverage
│   ├── id
│   │   ├── [id].mosdepth.global.dist.txt
│   │   ├── [id].mosdepth.summary.txt
│   │   ├── [id].per-base.bed.gz
│   │   └── [id].per-base.bed.gz.csi
│   └── strain
│       ├── [strain].mosdepth.global.dist.txt
│       ├── [strain].mosdepth.summary.txt
│       ├── [strain].per-base.bed.gz
│       └── [strain].per-base.bed.gz.csi
├── software_versions.txt
└── summary.txt
</code></pre>

<p>Most files should be obvious. A few are detailed below.</p>
<ul>
<li><strong>software_versions.txt</strong> - Outputs the fosftware versions used for every process (step) of the pipeline.</li>
<li><strong>summary.txt</strong> - Outputs a summary of the parameters used.</li>
<li><strong>aggregate</strong> - Stores data that has been aggregated across all strains or sequencing IDs. </li>
<li><strong>coverage</strong> - Contains coverage data at the strain or id level, presented in a variety of ways.</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../pipeline-concordance/" class="btn btn-neutral float-right" title="Concordance">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../pipeline-nil-ril/" class="btn btn-neutral" title="NIL / RIL"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/Andersenlab/dry-guide/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../pipeline-nil-ril/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../pipeline-concordance/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
